modules/apps/transfer/keeper/msg_server.go

```diff
@@ -59,3 +66,33 @@ func (k Keeper) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams)
 
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+// unwindHops unwinds the hops present in the tokens denomination and returns the message modified to reflect
+// the unwound path to take. It ensures only a single token must be present in the tokens list
+// and that the token is not native to the chain.
+func (k Keeper) unwindHops(ctx sdk.Context, msg *types.MsgTransfer) (*types.MsgTransfer, error) {
+	coins := msg.GetCoins()
+	if len(coins) > 1 {
+		return nil, errorsmod.Wrap(types.ErrInvalidForwarding, "cannot unwind more that one tokens")
+	}
+
+	token, err := k.tokenFromCoin(ctx, coins[0])
+	if err != nil {
+		return nil, err
+	}
+
+	if token.Denom.IsNative() {
+		return nil, errorsmod.Wrap(types.ErrInvalidForwarding, "cannot unwind a native token")
+	}
+	var unwindHops []types.Hop
+	for _, trace := range token.Denom.Trace[1:] {
+		unwindHops = append(unwindHops, types.Hop{PortId: trace.PortId, ChannelId: trace.ChannelId}) //nolint: gosimple
+	}
+
+	// Update message fields.
+	msg.SourcePort, msg.SourceChannel = token.Denom.Trace[0].PortId, token.Denom.Trace[0].ChannelId
```

Comment:
Since we are overriding here the `SourcePort` and `SourceChannel`, maybe it's possible to relax [the validation of these fields in `ValidateBasic`](https://github.com/cosmos/ibc-go/blob/35181d87c6aabd10a89346586b9f67cee21acd69/modules/apps/transfer/types/msgs.go#L70-L75)? If `unwind` is `true`, then skip the validation and allow them to be empty, for example.

Reply:
Done